drivers/ata8510/ata8510.c

```diff
@@ -0,0 +1,543 @@
+    return data[2];
+}
+
+static void _read_rx_fifo(ata8510_t *dev, uint8_t *buffer, uint8_t buffer_len)
+{
+    uint8_t command[ATA8510_DFIFO_RX_LENGTH +
+                    3] = { ATA8510_CMD_READ_RX_FIFO, buffer_len };
+    uint8_t data[ATA8510_DFIFO_RX_LENGTH + 3];
+
+    _send_command(dev, command, data, sizeof(command));
+    memcpy(buffer, data + 3, buffer_len);
+}
+
+static uint8_t _read_fill_level_rssi_fifo(ata8510_t *dev)
+{
+    uint8_t command[3] = { ATA8510_CMD_READ_FILL_LEVEL_RSSI_FIFO, 0x00, 0x00 };
+    uint8_t data[3];
+
+    _send_command(dev, command, data, sizeof(command));
+    return data[2];
+}
+
+static void _read_rssi_fifo(ata8510_t *dev, uint8_t *buffer, uint8_t buffer_len)
+{
+    uint8_t command[ATA8510_SFIFO_LENGTH +
+                    3] = { ATA8510_CMD_READ_RSSI_FIFO, buffer_len };
+    uint8_t data[ATA8510_SFIFO_LENGTH + 3];
+
+    _send_command(dev, command, data, sizeof(command));
+    memcpy(buffer, data + 3, buffer_len);
+}
+
+static int _recv_frame(ata8510_t *dev, uint8_t *buffer, uint8_t buffer_len,
+                       int *dbm)
+{
+    uint8_t *ptr = buffer;
+    uint8_t len = buffer_len;
+    uint8_t n, i;
+    uint8_t rssi_data[ATA8510_SFIFO_LENGTH];
+
+    while (len > 0) {
+        // check RX FIFO
+        int fifo = _read_fill_level_rx_fifo(dev);
+        if (fifo > 0) {
+            n = (fifo <= len ? fifo : len);
+            _read_rx_fifo(dev, ptr, n);
+            DEBUG("[ata8510] _recv_frame: read %d bytes\n", n);
+            ptr += n;
+            len -= n;
+        }
+        // check RSSI FIFO
+        fifo = _read_fill_level_rssi_fifo(dev);
+        if (fifo > 0) {
+            _read_rssi_fifo(dev, rssi_data, fifo);
+            for (i = 0; i < fifo; i++) {
+                dev->rssi.data[ (i + dev->rssi.len) %
+                                sizeof(dev->rssi.data) ] = rssi_data[i];
+            }
+            dev->rssi.len += fifo;
+            if (dev->rssi.len > sizeof(dev->rssi.data)) {
+                dev->rssi.len = sizeof(dev->rssi.data);
+            }
+        }
+        // if end of transfer, bail out
+        if (dev->events.events.val.eota) {
+            break;
+        }
+        if (len > 0) {
+            do {
+                mutex_lock(&(dev->rx_lock));
+                ata8510_get_event_bytes(dev);
+            } while (!dev->events.system.val.dfifo_rx &&
+                     !dev->events.events.val.eota &&
+                     !dev->events.system.val.sfifo);
+        }
+    }
+
+    if (!dev->events.events.val.eota) {
+        DEBUG(
+            "[ata8510] _recv_frame: ERROR - insufficient buffer, "
+            "packet truncated\n");
+    }
+
+    n = buffer_len - len;
+    DEBUG("[ata8510] _recv_frame: received %d bytes\n", n);
+
+    if (dev->rssi.len > 0) {
+        int rssi = 0;
+        for (i = 0; i < dev->rssi.len; i++) {
+            rssi += dev->rssi.data[i];
+        }
+        rssi /= dev->rssi.len;
+        *dbm = (rssi >> 1) - 135;
```

Comment:
Is that RSSI algorithm described in the data sheet?

Reply:
The conversion from RSSI to dBm,? Yes, it's taken form the datasheet.